### ISSUE TITLE
remote-plugin documentation confusion

### ISSUE THREAD

**Issue by osa1** [neovim/neovim#1929]
I'm confused about manifest files. Specifically, this paragraph:

> `:UpdateRemotePlugins` will generate the remote plugin manifest, a special
> vimscript file containing declarations for all vimscript entities
> (commands/autocommands/functions) defined by all remote plugins, with each
> entity associated with the host and plugin path. The manifest can be seen as a
> generated extension to the user's vimrc (it even has the vimrc filename
> prepended).

The documentation doesn't mention how can `:UpdateRemotePlugins` generate manifest file from a plugin written in some language other than Python. (I guess Python API provides a generation mechanism based on method decorators, but that's also not mentioned in the documentation)

It also doesn't give any example manifest files. (I think this is an example https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/host.vim#L58-63) and doesn't mention how nvim runs plugins written in any language. (e.g. starting the plugin program)

So in summary, I think the documentation needs to be improved by:

* Describing manifest file format, giving an example manifest file generated from/for a real API.
* Describing how plugins written in languages other than Python(or Python plugins written without using official Python API) are supposed to generate manifest files.
* Describing how to say nvim about running plugins implemented in languages other than Python.

Are those described in somewhere else? If someone could provide me some pointers I can update the documentation and send a pull request.

**Comment by justinmk** [comment]
I believe the current requirement is to modify this:

https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/host.vim#L2-L6

And add a corresponding file here: https://github.com/neovim/neovim/tree/master/runtime/autoload/provider

Sorry if that doesn't help much, there are some usability improvements I think we can make to rplugins.

**Comment by osa1** [comment]
Thanks for the answer, this was really helpful because I realized that I misunderstand the concept of "host".

> For AOT compiled languages, you should probably use the host factory function to compile discovered plugins into a single executable that accepts msgpack-rpc over stdio, only recompiling when one of the plugins is updated.

This would certainly work but I'm wondering if we could have something like this: Compile-to-native languages(i.e. Haskell, OCaml, Rust, Go, C++) don't need a shared host. It might be nice for Neovim to call plugins directly instead.

For example, Neovim could assume that compiled plugins support a command line argument to generate the API specification. Then it could directly load all plugins, without running a host etc.

If I understand the idea correctly, host idea is definitely good for plugins written in JVM and CLR languages(and probably some other interpreted languages) but for compiled-to-native languages it's making things harder.

What would be advantages and disadvantages of this approach? Could we implement/extend Neovim host registration API so that it allows this way of registering plugins?

I know I can always run arbitrary programs, and register some commands using Neovim RPC API so in theory I'm not bound to any hosts, but this way of loading plugins is a bit inconvenient for the users, since they need to be manually loaded(or some modifications in .vimrc is needed for something like `au FileType haskell <load the plugin>`).

**Comment by tarruda** [comment]
> This would certainly work but I'm wondering if we could have something like this: Compile-to-native languages(i.e. Haskell, OCaml, Rust, Go, C++) don't need a shared host. It might be nice for Neovim to call plugins directly instead.

There's nothing that binds plugins to use a shared host, in fact with some minor tweaking you can load multiple python plugins each in its own host.

> For example, Neovim could assume that compiled plugins support a command line argument to generate the API specification. Then it could directly load all plugins, without running a host etc.

The term "host" doesn't even exist in Neovim C code,  the remote plugin infrastructure is implemented in pure vimscript and is nothing but a simple convenience for defining commands, functions and autocommands implemented in other languages.

Nothing stops you from building a program that understands msgpack-rpc via stdio and connect it with neovim using `rpcstart`.

> If I understand the idea correctly, host idea is definitely good for plugins written in JVM and CLR languages(and probably some other interpreted languages) but for compiled-to-native languages it's making things harder.

Actually, I've taken some care to make the remote/host infrastructure play well with compiled languages. I haven't tested it yet due to the lack of working implementations, but  you should be able to use the host factory actually compile the discovered plugins into a single native executable(meaning the host is rebuilt every time a plugin changes). The idea is to eventually let plugins be modified and reloaded without leaving the editor, even for compiled languages.

The only difference between running remote plugins for compiled languages vs interpreted languages is that with interpreted languages the build/execution of the host all happen in the same process.

> What would be advantages and disadvantages of this approach?

There are two big disadvantages to using one process per plugin:

- Depending on the number of plugins, it will be very slow to load everything(if they are all loaded at once)
- Nvim may hit some OS limits such as the maximum number of child process(people could in theory have 100 plugins)

>  Could we implement/extend Neovim host registration API so that it allows this way of registering plugins?

I'm its not clear to me how you want the registration API to be, could you give a concrete example?

**Comment by Shougo** [comment]
> There are two big disadvantages to using one process per plugin:

Yes, I agree with you.  Google Chrome has same problem.

**Comment by Shougo** [comment]
> Nvim may hit some OS limits such as the maximum number of child process(people could in theory have 100 plugins)

In Mac OSX, the resource is very poor(I'm vimproc's experiment).
Windows is not created to communicate a lot of processes.

**Comment by tarruda** [comment]
> In Mac OSX, the resource is very poor(I'm vimproc's experiment).

:+1: The lua client used to leave zombie nvim processes which caused functional test failures after about 100 tests

**Comment by saep** [comment]
I'm working on a haskell plugin provider and osa1 just joined the project. At the current stage, we conceptually just generate one big executable that has all plugins baked in.

So, to make it work I just have to do something like this?

```viml
let host = 'myhost'
let chan = rpcstart("/path/to/executable")
let specs = rpcnotify(chan, ???)
remote#host#RegisterPlugin(host, ???, specs) 
```

However, I do not know if there is already a specification for what the question mark placeholder should be. `'api-info'` for the first seems reasonable to me. And for the second, a function similar to `remote#host#RegisterPlugin` might be needed as we do not have a path. Or we can just put the plugin name in there if the path is null. I am just brainstorming here, though.

**Comment by tarruda** [comment]
> However, I do not know if there is already a specification for what the question mark placeholder should be. 'api-info' for the first seems reasonable to me. And for the second, a function similar to remote#host#RegisterPlugin might be needed as we do not have a path. Or we can just put the plugin name in there if the path is null. I am just brainstorming here, though

`remote#host#RegisterPlugin` is a wrapper on top of other vimscript functions that define commands, functions and autocommands that automatically call msgpack-rpc clients using `rpcrequest`/`rpcnotify`. I recommend you to read these files to acquire a detailed understanding of how everything works:

- https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/define.vim (Helpers for defining commands/functions/autocommands)
- https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/host.vim (Host/plugin registration/setup. Plugin registration call the functions in define.vim)

Since you are managing plugin discovery yourself, there's no need to use `remote#host#RegisterPlugin` because its sole purpose is to associate a discovered remote plugin(implemented by a source file or module in another language) with a set of commands/functions/autocommands. Your best bet is to directly call the functions called by `remote#host#RegisterPlugin`. For example, say that `/path/to/executable` implements a msgpack-rpc method identified by "METHOD", and you want to create a vimscript Command that will call this msgpack-rpc method, here's how you could do it:

```vim
function! s:RequireHaskellHost(name)
  return rpcstart("/path/to/executable")
endfunction

call remote#host#Register('haskell', function('s:RequireHaskellHost'))
call remote#define#CommandOnHost('haskell', 'METHOD', 1, 'HaskellCommand', {}) 
```

What happens now is that the first time `HaskellCommand` is executed, it will call `s:RequireHaskellHost` to bootstrap the host and at the same time it will call 'METHOD'

**Comment by saep** [comment]
If I register a function via

```vimL
call remote#define#FunctionOnHost('haskell', 'METHOD', 1, 'Foo', {})
```
and then call

```vimL
call Foo(7)
```
The `7` is put in an extra list. Whereas calling the function via 

```vimL
rpcrequest(channel, 'METHOD', 7)
```
doesn't wrap the `7` in a list.

I know that this is caused by [this function](https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/define.vim#L158-L166). Specifically by putting `a:000` as the third argument to `rpcrequest` or `rpcnotify` respectively.

Is this behavior intended and should it stay this way?

**Comment by tarruda** [comment]
> Is this behavior intended and should it stay this way?

This is intended, without this wrapping it would be hard to distinguish arguments from other argument-like information such as ranges, bang or count.

**Comment by saep** [comment]
I think this issue can be closed now.

**Comment by vito-c** [comment]
```vim
function! s:RequireScalaHost(name)
    return rpcstart('/path/to/exec')
endfunction

call remote#host#Register('scala', "*.scala", function('s:RequireScalaHost'))
let sc=remote#host#Require('scala')
call remote#define#CommandOnHost('scala', 'METHOD', 1, 'ScalaCommand', {}) 
```

@saep I'm trying to do something similar in scala right now. When I use rpcstart('/path/to/exec') I get all sorts of weird errors. 
<img width="647" alt="screen shot 2016-02-11 at 9 37 39 am" src="https://cloud.githubusercontent.com/assets/820137/12984652/92a997a8-d0a3-11e5-93da-6f1f2a9c4e77.png">

I am able to messages to my client when exporting NVIM_LISTEN_ADDRESS=127.0.0.1:8000

**Comment by tarruda** [comment]
@vito-c I'm not sure what errors you are getting, you seem to have posted a screenshot of the import section of a scala file.

When you call through `NVIM_LISTEN_ADDRESS`, you are invoking msgpack-rpc methods defined in Nvim. The `remote#define#CommandOnHost` line  will define a command named "ScalaCommand" on nvim. When this command is called, it will spawn your scala plugin host and invoke "METHOD" through msgpack-rpc. Did you add a msgpack-rpc handler for "METHOD" on the scala side?

**Comment by vito-c** [comment]
@tarruda the screenshot is what happens when I open a scala file and the vimscript above runs. I can also replicate this error by calling just doing `call rpcstart('/path/to/exe')` There is definitely some sort of error going on. If you see the yellow j's those are a result of me pressing `j` after nvim opens. I'm not in normal mode or insert mode because hitting ESC does nothing I have to control-c out of nvim and then open a new term.

Maybe I'm using rpcstart incorrectly?? The only thing in my exe file is `sbt -Dsbt.log.noformat=true run &> test.log` which just runs my app. The app is setup to connect via 127.0.0.1 on 8000

When I try to call `ScalaCommand` in nvim by running `:call ScalaCommand()` I get `E117: Unknown function: ScalaCommand`

**Comment by vito-c** [comment]
I changed the define on the remote#define#CommandOnHost to be async. So now at least the yellow highlighted text doesn' t show up and neovim doesn't crash. However I'm still not seeing any logging on my client when I call the `:ScalaCommand` (the command is registered though). I am also tailing the `~/.nvimlog` 

It should be noted that I am connecting to a tcp socket. The exe only launches the plugin. How does neovim know the `scala` host is the one connecting to the socket?  I think what is happening is my client is launched and then connects to neovim (through tcp socket). Then I type `:ScalaCommand` in neovim and neovim sends a rpc message but not over the socket connection??
Thanks for the help.
